### PR TITLE
added dead letters queue support

### PR DESCRIPTION
**Pull request by mkorolyov**
Added dead letters queue support.
According to AWS console it is possible to configure dead letters queue for a dead letters queue, same we can do in elasticmq too.

### CODE REVIEW

File: core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala [hunk 1]
```diff
@@ -103,6 +103,15 @@ trait QueueActorMessageOps extends Logging {
         messageQueue += internalMessage
         None
       } else if (messagesById.contains(id.id)) {
+        // Putting the msg to dead letters queue if exists or delete.
+        if (internalMessage.receiveCount >= queueData.maxReceiveCount) {
+          logger.info(s"send message $internalMessage to dead letters actor $deadLettersActorRef")
+          deadLettersActorRef
+            .map(_ ! SendMessage(internalMessage.toNewMessageData))
+            .getOrElse(internalMessage.deliveryReceipt.foreach(dr => deleteMessage(DeliveryReceipt(dr))))
+          return None
```
Comment (adamw) [inline core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala]: [minor] maybe instead of the `return`, this could be an if-else. That way the `if` is an expression whose result is returned. Also this method is quite long - maybe a method could be extracted, e.g. `receiveMessage(id)`?

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala [hunk 2]
```diff
@@ -130,15 +130,15 @@ class QueueActorMsgOpsTest extends ActorTest with QueueManagerForEachTest with D
...
       // Then
-      lookupBeforeReceiving.flatMap(_.deliveryReceipt) should be (None)
+      lookupBeforeReceiving.flatMap(_.deliveryReceipt) should be(None)
 
       val receivedReceipt = received.flatMap(_.deliveryReceipt)
       val lookedUpReceipt = lookupAfterReceiving.flatMap(_.deliveryReceipt)
 
-      receivedReceipt.size  should be > (0)
-      lookedUpReceipt should be ('defined)
+      receivedReceipt.size should be > (0)
```
Comment (adamw) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: [minor] there's quite a lot of changes from reformatting which make it harder to spot the "real" changes

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala [hunk 3]
```diff
@@ -147,7 +147,7 @@ class QueueActorMsgOpsTest extends ActorTest with QueueManagerForEachTest with D
     val q1 = createQueueData("q1", MillisVisibilityTimeout(1L))
 
     for {
-      Right(queueActor) <- queueManagerActor ? CreateQueue(q1)
+      Right(queueActor) <- queueManagerActor ? CreateQueue(q1.copy(maxReceiveCount = 2))
```
Comment (adamw) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: is this change needed for the test to pass?

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala [hunk 4]
```diff
@@ -237,13 +238,13 @@ class QueueActorMsgOpsTest extends ActorTest with QueueManagerForEachTest with D
...
       // Then
-      lookupResult should be (None)
+      lookupResult should be(None)
     }
   }
 
   waitTest("msg statistics should be updated") {
     // Given
-    val q1 = createQueueData("q1", MillisVisibilityTimeout(1L))
+    val q1 = createQueueData("q1", MillisVisibilityTimeout(1L)).copy(maxReceiveCount = 2)
```
Comment (adamw) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: same here - is this needed?

File: rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala [hunk 5]
```diff
@@ -298,7 +299,9 @@ class AmazonJavaSdkTestSuite extends FunSuite with Matchers with BeforeAndAfter
   test("should block message for the visibility timeout duration") {
     // Given
     val queueUrl = client.createQueue(new CreateQueueRequest("testQueue1")
-      .withAttributes(Map(defaultVisibilityTimeoutAttribute -> "1"))).getQueueUrl
+      .withAttributes(
+        Map(defaultVisibilityTimeoutAttribute -> "1", defaultMaxReceiveCountAttribute -> "3")))
```
Comment (adamw) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: Does this need to be specified in every test? Shouldn't an undefined value work, meaning no DLQ support?

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala [hunk 6]
```diff
@@ -13,6 +13,7 @@ import CreateQueueDirectives._
 
 trait CreateQueueDirectives { this: ElasticMQDirectives with QueueURLModule with AttributesModule with SQSLimitsModule =>
 
+  // DeadLettersQueue currently not supported via http call due to current API limitations. Consider moving to JSON?
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala]: Aren't DLQ's configured via attributes passed during queue creation? The docs (http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_CreateQueue.html) mention `RedrivePolicy`.  Where's the JSON?

File: server/src/main/scala/org/elasticmq/server/ElasticMQServerConfig.scala [hunk 7]
```diff
@@ -57,20 +57,37 @@ class ElasticMQServerConfig(config: Config) {
   case class CreateQueue(name: String,
     defaultVisibilityTimeoutSeconds: Option[Long],
     delaySeconds: Option[Long],
-    receiveMessageWaitSeconds: Option[Long])
+    receiveMessageWaitSeconds: Option[Long],
+    deadLettersQueue: Option[CreateQueue] = None,
+    maxReceiveCount: Int = 1)
```
Comment (adamw) [inline server/src/main/scala/org/elasticmq/server/ElasticMQServerConfig.scala]: Aren't DLQ's optional? Then this should be optional too, no?

File: README.md [hunk 8]
```diff
@@ -116,14 +116,24 @@ The queues are specified in a custom configuration file. For example, create a `
 ````
 include classpath("application.conf")
 
-queues {
-    queue1 {
+queues = [
```
Comment (adamw) [inline README.md]: If possible, I would prefer not to change the format of the `queues` config, so that people already using this won't have to migrate.

Some note in the readme about DLQ support and how it works would be great as well.

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 2 above)
Comment (mkorolyov) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: sorry, my IDE reformatted code by itself, will try to revert such changes.

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 3 above)
Comment (mkorolyov) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: yes, bc now number if receives is count and msg will not be available if we reach limit.

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 4 above)
Comment (mkorolyov) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: yes, same

File: rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala (hunk 5 above)
Comment (mkorolyov) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: this param added only to small part of tests, i added it int he way like `defaultVisibilityTimeoutAttribute` was done.
if needed we can introduce some DSL to make code more readable.

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala (hunk 6 above)
Comment (mkorolyov) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala]: didn't get it, RedrivePolicy expects a string with json like `"{\"maxReceiveCount\":\"5\", \"deadLetterTargetArn\":\"dead_letters_queue1\"}"` ?

File: core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala [hunk 9]
```diff
@@ -103,6 +103,20 @@ trait QueueActorMessageOps extends Logging {
         messageQueue += internalMessage
         None
       } else if (messagesById.contains(id.id)) {
+        // Putting the msg to dead letters queue if exists
+        if (internalMessage.receiveCount >= queueData.maxReceiveCount) {
+          logger.info(s"send message $internalMessage to dead letters actor $deadLettersActorRef")
+          deadLettersActorRef
+            .map(_ ! SendMessage(internalMessage.toNewMessageData))
```
Comment (sbinq) [inline core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala]: `foreach` is more idiomatic for side effects btw

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 3 above)
Comment (adamw) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: but shouldn't there be no limit by default? As far as I understand, that's how SQS behaves

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala (hunk 6 above)
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala]: Ah, true! So the JSON is the value of the `RedrivePolicy` attribute. Then we should probably parse that :)

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 3 above)
Comment (mkorolyov) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: seems there is inconsistent in sqs docs:
Here it says that maxReceiveCount is not required - http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues-redrivepolicy.html
While here it should be set from 1 to 1000: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues-redrivepolicy.html

File: core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala (hunk 9 above)
Comment (mkorolyov) [inline core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala]: yeah, sure, didn't change it in last fix commit, will update

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 3 above)
Comment (adamw) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: I think you are looking at the docs of AWS CloudFormation, not SQS - maybe there's a typo in one of the CloudFormation docs. In the SQS docs (e.g. http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_CreateQueue.html), the `RedrivePolicy` is specified as *optional* (which it must be, to maintain backwards-compatibility).

This is also visible in the web interface:
http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html

File: core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala (hunk 1 above)
Comment (mkorolyov) [inline core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala]: agree, fixed

File: core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala (hunk 3 above)
Comment (mkorolyov) [inline core/src/test/scala/org/elasticmq/actor/QueueActorMsgOpsTest.scala]: agree, thanks, fixed

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala (hunk 6 above)
Comment (mkorolyov) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala]: done

File: server/src/main/scala/org/elasticmq/server/ElasticMQServerConfig.scala (hunk 7 above)
Comment (mkorolyov) [inline server/src/main/scala/org/elasticmq/server/ElasticMQServerConfig.scala]: agree, fixed

File: README.md (hunk 8 above)
Comment (mkorolyov) [inline README.md]: done

File: core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala (hunk 9 above)
Comment (mkorolyov) [inline core/src/main/scala/org/elasticmq/actor/queue/QueueActorMessageOps.scala]: done

File: rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala [hunk 10]
```diff
@@ -894,6 +895,26 @@ class AmazonJavaSdkTestSuite extends FunSuite with Matchers with BeforeAndAfter
...
 
+  test("should create queue with redrive policy.") {
+    // When
+    val queueUrl = client.createQueue(new CreateQueueRequest("testQueue1")
+      .withAttributes(Map(
+        defaultVisibilityTimeoutAttribute -> "1",
+        redrivePolicyAttribute ->
+          """
+            |{
+            |  "queueName":"dlq1",
```
Comment (adamw) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: Keeping compatibility with SQS, shouldn't this be called `deadLetterTargetArn`? Also, the SQS docs (I'm looking here: http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html), don't mention the ability to specify e.g. the `defaultVisibilityTimeout`.

Comment (mkorolyov) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: i noticed that some other params in create queue request also has different name than in real sqs, so i decided to use more readable name, or i am wrong?
defaultVisibilityTimeout not mentioned in docs, but i was able to set it and all other params from AWS console.

Comment (adamw) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: they do? which ones? I think they should be the same ... if the tested code is supposed to work the same with real and mock SQS :)

can you set the `defaultVisibilityTimeout` when defining a dead letter queue for an existing queue? Or only when creating a queue?

**Comment by mkorolyov** [comment]
@adamw 
> SQS requires the dead letters queue to be already created
didn't get it.
in AWS console there is a possibility to add dead letters queue to normal queue and you don't need to create dead letters queue manually before adding, it is created when you need automatically.

**Comment by adamw** [comment]
I'm not that familiar with the console, I'm rather following the API docs, which ElasticMQ is supposed to follow (after all, we are implementing a mock SQS API implementation, not a mock SQS console :) )

File: rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala (hunk 10 above)
Comment (mkorolyov) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: sorry, yes, all params has correct names according docs, will update new ones.

about `defaultVisibilityTimeout` both cases are valid. it can be updated at any time.

Comment (adamw) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: But if the dead letter queue has to be created before setting it as a DLQ, then why would you pass the visibility timeout here?

Anyway, thanks a lot for the changes! :)

**Comment by mkorolyov** [comment]
@adamw checked one more time, yes, you are 100% right, dead letters queues should be created by hands before using.

So i need to update http handler a little bit not to create dead letters queues and return err if specified queue in RedrivePolicy is not created.

Will fix in a few days.
Thanks!

File: rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala (hunk 10 above)
Comment (mkorolyov) [inline rest/rest-sqs-testing-amazon-java-sdk/src/test/scala/org/elasticmq/rest/sqs/AmazonJavaSdkTestSuite.scala]: yes, i was sure that dead letters queue should not be precreated, as it is wrong, i will remove all excess params from `RedrivePolicy` json.

**Comment by mkorolyov** [comment]
@adamw 
Hey, had some time todays morning, updated PR according yesterday discussion.

**Comment by mkorolyov** [comment]
@adamw when are you planning to make next release?

**Comment by adamw** [comment]
@mkorolyov just release 0.13.0. I made a minor change to the config format, hope you don't mind :)

**Comment by adamw** [comment]
Thanks a lot for your contribution!

**Comment by mkorolyov** [comment]
thanks for fast feedbacks! 
BRs

**Comment by joelrfcosta** [comment]
I'm working with version 0.13.2 `=== ElasticMQ server (0.13.2) started in 18185 ms ===` but I'm getting this message on console:
```
Ignored attribute "RedrivePolicy" (supported by SQS but not ElasticMQ)
```
Isn't `RedrivePolicy` functionality available since 0.13.0 @adamw?

**Comment by adamw** [comment]
@joelrfcosta Ah, it's only supported in create queue right now. Can you create a bug for that?
